File: src/sandbox.ts
/**
 * @file sandbox.ts
 * @description Experimental agent execution guardrail orchestrator
 * @author Nymrel / JalenBuilds LLC <contact@nymrel.com>
 * @license MIT
 */

import * as path from 'node:path';
import { spawn } from 'node:child_process';
import type {
  SandstormOptions,
  SandboxContext,
  SandboxResult,
  ExecOptions,
  ExecResult,
  SnapshotMetadata,
  RollbackResult,
  CommitResult,
  WorkspaceDiff,
  CommandInput,
} from './types.js';
import { CoWSnapshotManager } from './cow/index.js';
import { ZeroTrustProxy } from './proxy/index.js';
import { ExecutionLimiter } from './limiter/index.js';
import { AuditLogger, exportTimelineAscii, exportHtmlReportToFile } from './audit/index.js';

function definedProcessEnv(): Record<string, string> {
  const environment: Record<string, string> = {};
  for (const [key, value] of Object.entries(process.env)) {
    if (value !== undefined) environment[key] = value;
  }
  return environment;
}

/** Parse a portable command string for the default shell-free execution path. */
function parseCommandLine(command: string): [string, string[]] {
  const args: string[] = [];
  let current = '';
  let quote: '"' | "'" | null = null;
  let tokenStarted = false;

  for (let index = 0; index < command.length; index += 1) {
    const char = command[index]!;
    const next = command[index + 1];

    if (char === '\\' && next !== undefined && (next === '\\' || next === '"' || next === "'" || /\s/.test(next))) {
      current += next;
      tokenStarted = true;
      index += 1;
      continue;
    }

    if (quote) {
      if (char === quote) quote = null;
      else current += char;
      tokenStarted = true;
      continue;
    }

    if (char === '"' || char === "'") {
      quote = char;
      tokenStarted = true;
    } else if (/\s/.test(char)) {
      if (tokenStarted) {
        args.push(current);
        current = '';
        tokenStarted = false;
      }
    } else {
      current += char;
      tokenStarted = true;
    }
  }

  if (quote) throw new Error('Command contains an unterminated quote');
  if (tokenStarted) args.push(current);
  if (args.length === 0) throw new Error('Command must not be empty');

  return [args[0]!, args.slice(1)];
}

function formatCommand(command: CommandInput): string {
  if (typeof command === 'string') return command;
  if (command.length === 0) throw new Error('Command must not be empty');

  return command.map((argument) => {
    if (typeof argument !== 'string') throw new TypeError('Command arguments must be strings');
    return /^[a-zA-Z0-9_./:@%+=,-]+$/.test(argument) ? argument : JSON.stringify(argument);
  }).join(' ');
}

export class CommandExecutionError extends Error {
  public readonly command: string;
  public readonly execution: ExecResult;

  constructor(command: string, execution: ExecResult) {
    super(`Command exited with code ${execution.exitCode}: ${command}`);
    this.name = 'CommandExecutionError';
    this.command = command;
    this.execution = execution;
  }
}

export class Sandstorm {
  public readonly workspace: string;
  public readonly cow: CoWSnapshotManager;
  public readonly proxy: ZeroTrustProxy;
  public readonly limiter: ExecutionLimiter;
  public readonly audit: AuditLogger;
  public readonly options: Required<SandstormOptions>;

  constructor(options: SandstormOptions) {
    this.workspace = path.resolve(options.workspace);
    const sandstormDir = path.join(this.workspace, '.sandstorm');

    this.options = {
      workspace: this.workspace,
      allowDomains: options.allowDomains ?? [],
      blockDomains: options.blockDomains || [],
      maxSpendUsd: options.maxSpendUsd ?? Infinity,
      maxTokens: options.maxTokens ?? Infinity,
      maxSteps: options.maxSteps ?? 100,
      maxDurationMs: options.maxDurationMs ?? Infinity,
      autoRollbackOnError: options.autoRollbackOnError ?? true,
      detectRunawayLoops: options.detectRunawayLoops ?? true,
      loopThreshold: options.loopThreshold ?? 3,
      scanSecrets: options.scanSecrets ?? true,
      customSecretPatterns: options.customSecretPatterns || [],
      auditLogPath: options.auditLogPath || path.join(sandstormDir, 'audit.jsonl'),
      silent: options.silent ?? false,
    };

    this.cow = new CoWSnapshotManager(this.workspace);

    this.audit = new AuditLogger({
      logFilePath: this.options.auditLogPath,
      initialPayload: {
        workspace: this.workspace,
        allowDomains: this.options.allowDomains,
        autoRollback: this.options.autoRollbackOnError,
      },
    });

    this.proxy = new ZeroTrustProxy({
      allowedDomains: this.options.allowDomains,
      blockedDomains: this.options.blockDomains,
      scanPayloads: this.options.scanSecrets,
      customSecretPatterns: this.options.customSecretPatterns,
      onSecretDetected: (detection) => {
        this.audit.recordEvent('SECRET_BLOCKED', 'critical', {
          pattern: this.proxy.redactForAudit(detection.patternName),
          severity: detection.severity,
          location: detection.location,
          redacted: detection.redactedText,
        });
      },
      onBlockedDomain: (domain, url) => {
        this.audit.recordEvent('DOMAIN_BLOCKED', 'warn', {
          domain: this.proxy.redactForAudit(domain),
          url: this.proxy.redactForAudit(url),
        });
      },
    });

    this.limiter = new ExecutionLimiter({
      maxSpendUsd: this.options.maxSpendUsd,
      maxTotalTokens: this.options.maxTokens,
      maxSteps: this.options.maxSteps,
      maxDurationMs: this.options.maxDurationMs,
      loopDetection: this.options.detectRunawayLoops,
      loopThreshold: this.options.loopThreshold,
    });
  }

  /**
   * Run a callback with the configured cooperative guardrails
   */
  public async run<T = unknown>(
    fn: (ctx: SandboxContext) => Promise<T>
  ): Promise<SandboxResult<T>> {
    const startTime = Date.now();
    this.limiter.reset();

    // 1. Create Pristine Baseline CoW Snapshot
    const baseSnapshot = this.cow.createSnapshot('pre-execution-baseline');
    this.audit.recordEvent('SNAPSHOT_CREATED', 'info', {
      snapshotId: baseSnapshot.id,
      name: baseSnapshot.name,
      fileCount: baseSnapshot.fileCount,
      treeHash: baseSnapshot.treeHash,
    });

    // 2. Start the cooperative outbound proxy
    await this.proxy.start();
    const proxyEnv = this.proxy.getEnv();

    let resultValue: T | undefined;
    let executionError: Error | undefined;
    let rollbackPerformed = false;
    let rollbackSummary: RollbackResult | undefined;

    const ctx: SandboxContext = {
      workspace: this.workspace,
      env: { ...definedProcessEnv(), ...proxyEnv },
      exec: async (command: CommandInput, options: ExecOptions = {}) => {
        const displayCommand = formatCommand(command);
        const actionName = typeof command === 'string' ? command.trim().split(/\s+/, 1)[0] : command[0];
        if (!actionName) throw new Error('Command must not be empty');
        this.limiter.recordStep(actionName, displayCommand);
        const redactedCommand = this.proxy.redactForAudit(displayCommand);
        this.audit.recordEvent('EXEC_STARTED', 'info', { command: redactedCommand });

        const execRes = await this.execCommandInternal(command, {
          ...options,
          cwd: options.cwd ?? this.workspace,
          // A caller may add variables, but must not silently replace the
          // proxy variables that define the guarded child-process boundary.
          env: { ...options.env, ...proxyEnv },
        });

        this.audit.recordEvent('EXEC_FINISHED', execRes.exitCode === 0 ? 'info' : 'warn', {
          command: redactedCommand,
          exitCode: execRes.exitCode,
          durationMs: execRes.durationMs,
        });

        if (execRes.exitCode !== 0 && !options.allowNonZeroExit) {
          throw new CommandExecutionError(redactedCommand, {
            ...execRes,
            stdout: this.proxy.redactForAudit(execRes.stdout),
            stderr: this.proxy.redactForAudit(execRes.stderr),
          });
        }

        return execRes;
      },
      recordTokenUsage: (model: string, promptTokens: number, completionTokens: number) => {
        const usage = this.limiter.recordTokens(model, promptTokens, completionTokens);
        this.audit.recordEvent('BUDGET_ACCUMULATED', 'info', {
          model: this.proxy.redactForAudit(model),
          promptTokens,
          completionTokens,
          spendUsd: usage.currentSpendUsd,
          totalTokens: usage.totalTokens,
        });
        return usage;
      },
      recordStep: (actionName: string, detail?: string) => {
        this.limiter.recordStep(actionName, detail);
        this.audit.recordEvent('STEP_EXECUTED', 'info', {
          actionName: this.proxy.redactForAudit(actionName),
          detail: detail === undefined ? undefined : this.proxy.redactForAudit(detail),
        });
      },
      snapshot: (name?: string) => {
        const snap = this.cow.createSnapshot(name);
        this.audit.recordEvent('SNAPSHOT_CREATED', 'info', {
          snapshotId: snap.id,
          name: snap.name,
          treeHash: snap.treeHash,
        });
        return snap;
      },
      rollback: (snapshotId?: string) => {
        const rb = this.cow.rollback(snapshotId);
        this.audit.recordEvent('ROLLBACK_TRIGGERED', 'warn', {
          snapshotId: rb.snapshotId,
          restored: rb.restoredFiles.length,
          deleted: rb.deletedFiles.length,
          reverted: rb.revertedFiles.length,
        });
        return rb;
      },
      diff: () => this.cow.diff(),
    };

    try {
      resultValue = await fn(ctx);
    } catch (err: unknown) {
      executionError = err instanceof Error ? err : new Error(String(err));
      this.audit.recordEvent('CIRCUIT_BREAKER_TRIPPED', 'critical', {
        error: this.proxy.redactForAudit(executionError.message),
        stack: executionError.stack === undefined ? undefined : this.proxy.redactForAudit(executionError.stack),
      });

      if (this.options.autoRollbackOnError) {
        rollbackSummary = this.cow.rollback(baseSnapshot.id);
        rollbackPerformed = true;
        this.audit.recordEvent('ROLLBACK_TRIGGERED', 'warn', {
          snapshotId: baseSnapshot.id,
          success: rollbackSummary.success,
          restored: rollbackSummary.restoredFiles.length,
          deleted: rollbackSummary.deletedFiles.length,
          reverted: rollbackSummary.revertedFiles.length,
          durationMs: rollbackSummary.durationMs,
          error: rollbackSummary.error,
        });
      }
    } finally {
      await this.proxy.stop();
    }

    const spendSummary = this.limiter.getSummary();
    const auditVerified = this.audit.verifyIntegrity();
    const durationMs = Date.now() - startTime;
    const finalTreeHash = this.cow.createSnapshot('final-state').treeHash;

    return {
      success: !executionError,
      result: resultValue,
      error: executionError,
      rollbackPerformed,
      rollbackSummary,
      baseSnapshot,
      finalTreeHash,
      auditSummary: {
        eventCount: this.audit.getEvents().length,
        verified: auditVerified.valid,
        latestHash: this.audit.getLatestHash(),
      },
      spendSummary: {
        totalSpendUsd: spendSummary.totalSpendUsd,
        totalTokens: spendSummary.totalTokens,
        stepsExecuted: spendSummary.stepsExecuted,
      },
      durationMs,
    };
  }

  /**
   * Helper to execute a command with sandboxed environment variables
   */
  public async exec(command: CommandInput, options: ExecOptions = {}): Promise<ExecResult> {
    return this.execCommandInternal(command, {
      cwd: this.workspace,
      ...options,
    });
  }

  private execCommandInternal(command: CommandInput, options: ExecOptions = {}): Promise<ExecResult> {
    return new Promise((resolve) => {
      const startTime = Date.now();
      const spawnOptions = {
        cwd: options.cwd || this.workspace,
        env: { ...process.env, ...options.env },
      };
      const shell = options.shell ?? false;
      if (shell && typeof command !== 'string') {
        throw new TypeError('Shell execution requires a command string');
      }
      const [executable, args] = typeof command === 'string'
        ? parseCommandLine(command)
        : (() => {
            if (command.length === 0) throw new Error('Command must not be empty');
            if (command.some((argument) => typeof argument !== 'string')) {
              throw new TypeError('Command arguments must be strings');
            }
            return [command[0]!, [...command.slice(1)]] as [string, string[]];
          })();
      const child = shell
        ? spawn(command as string, { ...spawnOptions, shell })
        : spawn(executable, args, { ...spawnOptions, shell: false });

      let stdout = '';
      let stderr = '';

      child.stdout?.on('data', (chunk) => {
        stdout += chunk.toString();
      });

      child.stderr?.on('data', (chunk) => {
        stderr += chunk.toString();
      });

      let timer: NodeJS.Timeout | null = null;
      if (options.timeoutMs) {
        timer = setTimeout(() => {
          child.kill('SIGKILL');
          stderr += `\nCommand timed out after ${options.timeoutMs}ms`;
        }, options.timeoutMs);
      }

      child.on('close', (exitCode) => {
        if (timer) clearTimeout(timer);
        resolve({
          // A process closed by a signal has no numeric exit code. Treat that
          // as failure; mapping null to zero would incorrectly report success.
          exitCode: exitCode ?? 1,
          stdout,
          stderr,
          durationMs: Date.now() - startTime,
        });
      });

      child.on('error', (err) => {
        if (timer) clearTimeout(timer);
        resolve({
          exitCode: 1,
          stdout,
          stderr: `${stderr}\n${err.message}`,
          durationMs: Date.now() - startTime,
        });
      });
    });
  }

  public snapshot(name?: string): SnapshotMetadata {
    return this.cow.createSnapshot(name);
  }

  public rollback(snapshotId?: string): RollbackResult {
    return this.cow.rollback(snapshotId);
  }

  public diff(snapshotId?: string): WorkspaceDiff {
    return this.cow.diff(snapshotId);
  }

  public commit(name?: string): CommitResult {
    return this.cow.commit(name);
  }

  public getTimeline(): string {
    return exportTimelineAscii(this.audit.getEvents());
  }

  public exportHtmlReport(filePath: string): void {
    const integrity = this.audit.verifyIntegrity();
    const spend = this.limiter.getSummary();
    exportHtmlReportToFile(filePath, this.audit.getEvents(), integrity, {
      workspace: this.workspace,
      totalSpendUsd: spend.totalSpendUsd,
      totalTokens: spend.totalTokens,
    });
  }
}

File: python/agent_sandstorm/sandbox.py
"""
sandbox.py: Master Sandstorm Orchestrator for Python
Copyright 2026 Nymrel / JalenBuilds LLC <contact@nymrel.com>
MIT License
"""

import os
import shlex
import time
import subprocess
from dataclasses import dataclass, field
from typing import Optional, List, Dict, Callable, Any

from .cow import CoWSnapshotManager, Snapshot, RollbackResult
from .proxy import ZeroTrustProxy
from .limiter import ExecutionLimiter
from .audit import AuditLogger, export_timeline_ascii, generate_html_report


@dataclass
class SandboxResult:
    success: bool
    result: Any = None
    error: Optional[Exception] = None
    rollback_performed: bool = False
    rollback_summary: Optional[RollbackResult] = None
    base_snapshot: Optional[Snapshot] = None
    final_tree_hash: str = ""
    duration_ms: float = 0.0
    audit_summary: Dict[str, Any] = field(default_factory=dict)
    spend_summary: Dict[str, Any] = field(default_factory=dict)


class CommandExecutionError(RuntimeError):
    """Raised when a guarded child command exits unsuccessfully."""

    def __init__(self, command: str, execution: subprocess.CompletedProcess):
        super().__init__(f"Command exited with code {execution.returncode}: {command}")
        self.command = command
        self.execution = execution


class Sandstorm:
    def __init__(
        self,
        workspace: str,
        allow_domains: Optional[List[str]] = None,
        block_domains: Optional[List[str]] = None,
        allow_ports: Optional[List[int]] = None,
        max_spend_usd: Optional[float] = None,
        max_tokens: Optional[int] = None,
        max_steps: int = 100,
        max_duration_ms: Optional[float] = None,
        auto_rollback_on_error: bool = True,
        detect_runaway_loops: bool = True,
        scan_secrets: bool = True,
        audit_log_path: Optional[str] = None,
    ):
        self.workspace = os.path.abspath(workspace)
        self.auto_rollback = auto_rollback_on_error
        self.cow = CoWSnapshotManager(self.workspace)

        if not audit_log_path:
            audit_log_path = os.path.join(self.workspace, ".sandstorm", "audit.jsonl")

        self.audit = AuditLogger(log_file_path=audit_log_path)
        self.limiter = ExecutionLimiter(
            max_spend_usd=max_spend_usd,
            max_total_tokens=max_tokens,
            max_steps=max_steps,
            max_duration_ms=max_duration_ms,
            loop_detection=detect_runaway_loops,
        )

        allowed = allow_domains if allow_domains is not None else []
        self.proxy = ZeroTrustProxy(
            allowed_domains=allowed,
            blocked_domains=block_domains,
            allowed_ports=allow_ports,
            scan_payloads=scan_secrets,
            on_secret_detected=lambda d: self.audit.record_event(
                "SECRET_BLOCKED", "critical", {"pattern": d.pattern_name, "redacted": d.redacted_text}
            ),
            on_blocked_port=lambda host, port, url: self.audit.record_event(
                "PORT_BLOCKED",
                "warn",
                {
                    "host": self.proxy.scanner.redact_all(host),
                    "port": "invalid" if port is None else port,
                    "url": self.proxy.scanner.redact_all(url),
                },
            ),
        )

    def run(self, fn: Callable[[Any], Any]) -> SandboxResult:
        start_time = time.time()
        self.limiter.reset()
        base_snap = self.cow.create_snapshot("pre-execution-baseline")
        self.audit.record_event("SNAPSHOT_CREATED", "info", {"id": base_snap.id, "tree_hash": base_snap.tree_hash})

        self.proxy.start()
        result_val = None
        exec_err: Optional[Exception] = None
        rollback_done = False
        rb_summary: Optional[RollbackResult] = None

        class Context:
            def __init__(ctx_self):
                ctx_self.workspace = self.workspace
                ctx_self.env = {**os.environ, **self.proxy.get_env()}

            def exec(ctx_self, command, allow_nonzero: bool = False, shell: bool = False) -> subprocess.CompletedProcess:
                if isinstance(command, str):
                    display_command = command
                    invocation = command if shell else shlex.split(command, posix=os.name != "nt")
                else:
                    if shell:
                        raise TypeError("shell=True requires a command string")
                    invocation = [str(part) for part in command]
                    display_command = shlex.join(invocation)

                if not invocation:
                    raise ValueError("command must not be empty")

                action_name = invocation.split()[0] if isinstance(invocation, str) else invocation[0]
                self.limiter.record_step(action_name, display_command)
                redacted_command = self.proxy.scanner.redact_all(display_command)
                self.audit.record_event("EXEC_STARTED", "info", {"cmd": redacted_command, "shell": shell})
                # shell=True is permitted only through the explicit caller opt-in above.
                res = subprocess.run(
                    invocation,
                    shell=shell,  # nosec B602
                    cwd=self.workspace,
                    env=ctx_self.env,
                    capture_output=True,
                    text=True,
                )
                self.audit.record_event("EXEC_FINISHED", "info" if res.returncode == 0 else "warn", {"code": res.returncode})
                if res.returncode != 0 and not allow_nonzero:
                    redacted_execution = subprocess.CompletedProcess(
                        args=redacted_command,
                        returncode=res.returncode,
                        stdout=self.proxy.scanner.redact_all(res.stdout),
                        stderr=self.proxy.scanner.redact_all(res.stderr),
                    )
                    raise CommandExecutionError(redacted_command, redacted_execution)
                return res

            def record_tokens(ctx_self, model: str, p_tok: int, c_tok: int):
                usage = self.limiter.record_tokens(model, p_tok, c_tok)
                self.audit.record_event(
                    "BUDGET_ACCUMULATED",
                    "info",
                    {**usage, "model": self.proxy.scanner.redact_all(model)},
                )
                return usage

        ctx = Context()

        try:
            result_val = fn(ctx)
        except Exception as e:
            exec_err = e
            self.audit.record_event(
                "CIRCUIT_BREAKER_TRIPPED",
                "critical",
                {"error": self.proxy.scanner.redact_all(str(e))},
            )
            if self.auto_rollback:
                rb_summary = self.cow.rollback(base_snap.id)
                rollback_done = True
                self.audit.record_event(
                    "ROLLBACK_TRIGGERED",
                    "warn",
                    {
                        "success": rb_summary.success,
                        "restored": len(rb_summary.restored_files),
                        "deleted": len(rb_summary.deleted_files),
                        "reverted": len(rb_summary.reverted_files),
                        "duration_ms": rb_summary.duration_ms,
                        "error": rb_summary.error,
                    },
                )
        finally:
            self.proxy.stop()

        duration_ms = (time.time() - start_time) * 1000.0
        integrity = self.audit.verify_integrity()
        final_tree_hash = self.cow.create_snapshot("final-state").tree_hash

        return SandboxResult(
            success=exec_err is None,
            result=result_val,
            error=exec_err,
            rollback_performed=rollback_done,
            rollback_summary=rb_summary,
            base_snapshot=base_snap,
            final_tree_hash=final_tree_hash,
            duration_ms=duration_ms,
            audit_summary={"verified": integrity["valid"], "events": len(self.audit.events)},
            spend_summary=self.limiter.get_summary(),
        )

    def snapshot(self, name: Optional[str] = None) -> Snapshot:
        return self.cow.create_snapshot(name)

    def rollback(self, snapshot_id: Optional[str] = None) -> RollbackResult:
        return self.cow.rollback(snapshot_id)

    def diff(self, snapshot_id: Optional[str] = None):
        return self.cow.diff(snapshot_id)

    def commit(self, name: Optional[str] = None) -> Snapshot:
        return self.cow.commit(name)

    def get_timeline(self) -> str:
        return export_timeline_ascii(self.audit.events)
